DAOS-5507 daos: add serial daos filesystem copy

dsikich opened the pull request:
A filesystem copy allows copies to and from POSIX for DAOS->POSIX, POSIX->DAOS, and DAOS->DAOS. This version was created as a serial daos utility, which does not require MPI. This version uses many of the same abstractions that I put into mpifileutils dcp, such as many of the I/O wrappers that check if the file type is DAOS or POSIX. I'm not sure if where I placed the I/O wrappers (src/utils/daos_hdlr.c) is the best place for them, so I can move them if necessary.

dsikich commented:
@mchaarawi @daltonbohning this is my first DAOS PR so I opened it as a draft. I want to make sure the code is styled and organized correctly based on how DAOS does things first.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	fname = basename(f1);
+	cont_name = dirname(f2);
+
+	if (cont_name[0] == '.' || cont_name[0] != '/') {
```
```suggestion
	if (cont_name[0] != '/') {
```
`== '.'` is a subset of `!= '/'`, so checking for `'.'` is redundant

(I've also seen this function in other places, not including this PR)

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		rc = dfs_lookup(daos_file->dfs, dname, O_RDWR, &parent, NULL, NULL);
+		if (parent == NULL)
+			fprintf(stderr, "dfs_lookup %s failed \n", dname);
+		rc = dfs_mkdir(daos_file->dfs, parent, name, S_IRWXU, 0);
```
I wonder if we need a way to pass `mode` here, instead of always using `S_IRWXU`?

dsikich replied:
@daltonbohning yeah, not sure. I could pass in the mode from the source directory I suppose.

daltonbohning replied:
Since dfs doesn't do access checks on the permissions - it just stores them - I think this could work for a dfs destination.
But in the general case, such as POSIX, we can't immediately pass the permissions from the source, because if the source doesn't have write permission, then the destination will not either.
mpifileutils handles this by updating all permissions after copying.
I'll have to give this some more thought...

dsikich replied:
@daltonbohning I do think this is something to think about. But if the source doesn't have write permission, would we want the destination to have it?

daltonbohning replied:
Right, we _**don't**_ want the destination to have write permission if the source doesn't have it.
But for POSIX, we have to set write permission on the destination in order to write to the file. Then, _**after**_ copying, we can set the destination permissions to match the source exactly.
Of course, for a destination DAOS container we can just set the permissions immediately since DFS doesn't check the access - it just stores the POSIX permissions.

So, in a general case, we can't call, say, `daos_file_mkdir` with the source permissions. This works for DAOS but not for POSIX.

dsikich replied:
@daltonbohning ah yeah I see what you mean. We can do something similar.

daltonbohning replied:
@dsikich If we aren't yet going to handle this as discussed, would you please leave a TODO comment or something similar? There aren't yet datamover tests for verifying that permissions are correct, so I'm worried this might slip through accidentally.

dsikich replied:
@daltonbohning yes, definitely. I added one.

daltonbohning replied:
Thanks!

mchaarawi replied:
well it should be the same as the file create function, so we have to add a mode parameter from the calling function

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		const char* dir_name,
+		const char* fs_dst_prefix)
+{
+	DIR *d;
```
Maybe:
```suggestion
	DIR *src_dir;
```
Or:
```suggestion
	DIR *src_d;
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	int rc;
+
+	/* begin by opening source directory */
+	d = daos_file_opendir(daos_src_file, dir_name);
```
To be clear, we are only allowing a source directory, not a source file?

dsikich replied:
@daltonbohning yes, in the case that it moving data from a DAOS container to a source file I am trying to think about if that would make sense. The DAOS source would need to be a single file, which is a case we should probably handle.

daltonbohning replied:
I think we probably should, or at least error. But maybe we can do that in a future iteration?

dsikich replied:
@daltonbohning yeah, for now we can error. This PR is already fairly large so I can add support for that in the next iteration.

daltonbohning replied:
I agree. And once we have a test plan and start implementing tests we can hammer out some of the finer points like this.

dsikich replied:
@daltonbohning ok, I've added an error check here if the source is not a directory, we can also work out better in which cases we want to return from this function

daltonbohning replied:
> which cases we want to return from this function

Do you mean to do this on this PR, or future? I'm okay with a future PR, since we have some other finer points to handle as well.

dsikich replied:
@daltonbohning in another PR, thanks for all of your feedback

daltonbohning replied:
You're welcome!

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				goto out;
+			}
+
+			strcpy(new_dir, cwd);
+			if (cont_name[0] == '.') {
+				strcat(new_dir, &cont_name[1]);
+			} else {
+				strcat(new_dir, "/");
+				strcat(new_dir, cont_name);
```
I wonder if we can utilize `snprintf` here for the `strcpy` and `strcat` for performance. (I have not tested)

(Again, this similar function exists elsewhere)

dsikich replied:
@daltonbohning yes, this parse_filename function is also in mpifileutils. I think that one still uses the strcpy and strcat as well though.

daltonbohning replied:
Yeah, it does. I think we can probably leave this function as-is for now, until we figure out how we're going to centralize the code. I believe this function in mpifileutils was also copied from somewhere else, so we have at least 3 copies/verisons that I''m aware of.

mchaarawi replied:
more than that :-)
i've written this function originally for the mpiio driver in mpich. now it's also in ior and HDF5.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+
+	assert(dir_name);
+
+	rc = dfs_lookup(daos_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
```
In a future iteration, we might consider using a `d_hash` to cache some lookups

dsikich replied:
@daltonbohning yep it would be nice if we can get to where we can just re-use the I/O wrappers across DAOS and mpifileutils

mchaarawi replied:
i think performance wise, it's pretty important to add a hashtable there. 
i've done something similar in the ior DFS driver.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+			uint64_t total_bytes = st.st_size;
+			uint64_t left_to_read = 0;
+			uint64_t buf_size = 64*1024*1024;
+			void* buf = malloc(buf_size * sizeof(char));
```
```suggestion
			D_ALLOC(buf, buf_size * sizeof(char));
				if (buf == NULL)
					return ENOMEM;
```
I think we can use `D_ALLOC`, and also we need to check for `ENOMEM`

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				left_to_read += (uint64_t) bytes_read;
+			}
+			if (buf != NULL)
+				free(buf);
```
```suggestion
				D_FREE(buf);
```
To match `D_ALLOC`

daltonbohning commented on src/utils/daos_hdlr.h:
```diff
@@ -69,11 +92,22 @@ struct cmd_args_s {
 	uuid_t			p_uuid;		/* --pool */
+	uuid_t			src_p_uuid;	/* --src_pool */
+	uuid_t			dst_p_uuid;	/* --dst_pool */
+	uuid_t			src_cont_uuid;	/* --src_cont */
+	uuid_t			dst_cont_uuid;	/* --dst_cont */
```
```suggestion
	uuid_t			src_c_uuid;	/* --src_cont */
	uuid_t			dst_c_uuid;	/* --dst_cont */
```

While I do prefer `cont`, the existing `--cont` using `c`.
Maybe we should be consistent?

daltonbohning commented on src/utils/daos_hdlr.h:
```diff
@@ -69,11 +92,22 @@ struct cmd_args_s {
+	uuid_t			dst_p_uuid;	/* --dst_pool */
+	uuid_t			src_cont_uuid;	/* --src_cont */
+	uuid_t			dst_cont_uuid;	/* --dst_cont */
+	char			*src_svc_str;	/* --src_svc */
+	d_rank_list_t		*src_svc;
+	char			*dst_svc_str;	/* --dst_svc */
+	d_rank_list_t		*dst_svc;
```
```suggestion
	char			*src_mdsrv_str;	/* --src_svc */
	d_rank_list_t		*src_mdsrv;
	char			*dst_mdsrv_str;	/* --dst_svc */
	d_rank_list_t		*dst_mdsrv;
```
Again, consistency?

daltonbohning commented on src/utils/daos.c:
```diff
@@ -925,6 +1012,23 @@ call_dfuse_ioctl(char *path, struct dfuse_il_reply *reply)
+	op = ap->fs_op;
+	switch (op) {
+	case FS_COPY:
+		fs_copy_hdlr(ap);
```
I'm not familiar with the process here, but do we need an error code from `fs_copy`?

daltonbohning replied:
I think this then brings to discussion which errors in `fs_copy` should return, and which should just be logged and continue.
I'm not sure about each one... But something to think about

daltonbohning reviewed (commented):
There is a lot here (good job) so I didn't go through every line in detail, but from a overview of the code, it looks mostly solid.
Before this is merged, I can run some of the dcp copy tests (adjusted for this of course) as sanity checks.
I'm also tasked with creating a test plan, so thorough testing will be done at that time.

In general, it looks like a lot of the I/O calls we have here are similar to what we have in mpifileutils. @dsikich It might be a good idea for use to discuss how we might potentially centralize this code. There are some caveats in some of the I/O calls that I've already addressed in mpifileutils, as well as ongoing changes - mostly in the realm of caching and handling symlinks. Centralizing the code would make sure this tool works similar to mpifileutils, reduce future maintenance, and allow us to potentially move more code from mpifileutils to our DAOS internals. @mchaarawi Might have some ideas as well, such as creating some sort of `dfs_wrapper` . Anyway, we definitely don't need to address that on this PR, but something to think about.

dsikich commented:
> There is a lot here (good job) so I didn't go through every line in detail, but from a overview of the code, it looks mostly solid.
> Before this is merged, I can run some of the dcp copy tests (adjusted for this of course) as sanity checks.
> I'm also tasked with creating a test plan, so thorough testing will be done at that time.
> 
> In general, it looks like a lot of the I/O calls we have here are similar to what we have in mpifileutils. @dsikich It might be a good idea for use to discuss how we might potentially centralize this code. There are some caveats in some of the I/O calls that I've already addressed in mpifileutils, as well as ongoing changes - mostly in the realm of caching and handling symlinks. Centralizing the code would make sure this tool works similar to mpifileutils, reduce future maintenance, and allow us to potentially move more code from mpifileutils to our DAOS internals. @mchaarawi Might have some ideas as well, such as creating some sort of `dfs_wrapper` . Anyway, we definitely don't need to address that on this PR, but something to think about.

@daltonbohning I agree, I had to re-use a lot of similar code from dcp, and it would be ideal to not have to duplicate so much across mpifileutils and DAOS. I am not sure the best approach for that, a dfs_wrapper or something could work, maybe we can discuss this in the next POSIX meeting. Especially since we will have a similar issue for the serialization code as well.

daltonbohning commented:
> @daltonbohning I agree, I had to re-use a lot of similar code from dcp, and it would be ideal to not have to duplicate so much across mpifileutils and DAOS. I am not sure the best approach for that, a dfs_wrapper or something could work, maybe we can discuss this in the next POSIX meeting. Especially since we will have a similar issue for the serialization code as well.

I'll make a note to bring this up in the next meeting.

daltonbohning reviewed (commented):
Something else I realized is that in mpifileutils, many of the posix I/O calls are wrapped with retries for `EINTR` and `EIO`, which these POSIX calls do not have.
I think it could be possible to define a macro in such a way that we can wrap most of the calls with this retry.

daltonbohning commented:
FYI: https://jira.hpdd.intel.com/browse/DAOS-4433
This is in progress and aims to remove svcl params entirely.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,758 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+			   &src_cont_info, &dst_cont_info,
+			   &src_dattr, &dst_dattr);
+
+	/* set paths based on file type for source and destination */
```
We don't handle POSIX -> POSIX here, and I imagine that's probably okay, but do we want to handle it, or at least have an `else ... error`?

daltonbohning replied:
@dsikich Actually, I think we might want to handle POSIX -> POSIX, for the case of DFUSE -> DFUSE

dsikich replied:
@daltonbohning I don't really see why someone would do that? If they are using two dfuse mounted directories why not just use cp?

dsikich replied:
@daltonbohning it is just kind of an odd thing to do, but I guess we can account for this. Right now, I will add a message that says it is not supported now at the very least.

daltonbohning replied:
@dsikich Yeah, it would be weird, but I imagine if someone _**did**_ try, they might be confused about why they can't do it. Since we support POSIX as the source or as the destination, I don't think it should be difficult to allow that option, right?
But yeah, we don't have to add it right now

daltonbohning replied:
Also, I think the POSIX->POSIX case would allow us to create a baseline to compare against `cp`. I.e. we can run both `cp` and `daos fs copy` on edge cases to make sure they are consistent. Then, as long as the DFS I/O functions match the POSIX I/O functions, we can know that we're handling these edge cases correctly.
I've actually done this when testing `dcp`.

dsikich replied:
@daltonbohning yeah shouldn't be hard to add

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,758 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+
+	/* stat the source, and make sure it is a directory  */
+	struct stat st_dir_name;
+	rc = daos_file_stat(daos_src_file, dir_name, &st_dir_name);
+	if (!S_ISDIR(st_dir_name.st_mode))
+		fprintf(stderr, "Source is not a directory: %s\n", dir_name);
+
```
@daltonbohning I am not entirely sure either, I have seen fprintf(stderrr,..) used as well as D_ERROR I wasn't sure about that either. I'm pretty new to the DAOS code base, so I tend to just do what I see in other code in the file or if it is mentioned in the style guide.

daltonbohning replied:
@dsikich Oh... You're right. Then I'm not sure if we should use D_ERROR or printf here

daltonbohning commented on src/utils/daos.c:
```diff
@@ -1280,6 +1384,19 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 
 	if (argc <= 2) {
 		FIRST_LEVEL_HELP();
+	} else if (strcmp(argv[2], "filesystem") != 0) {
+		if (strcmp(argv[3], "copy") == 0) {
```
The equality is reversed, we need to protect against the case where another argument isn't passed at all, and we should probably handle the shorthand "fs". 
Consider these cases:
`daos help filesystem`
`daos help fs`
```suggestion
	} else if (strcmp(argv[2], "filesystem") == 0 || strcmp(argv[2], "fs") == 0) {
		if (argc == 3) {
            		ALL_FS_CMDS_HELP();
		} else if (strcmp(argv[3], "copy") == 0) {
```

daltonbohning commented on src/utils/daos.c:
```diff
@@ -45,15 +45,25 @@
+{
+	if (strcmp(str, "copy") == 0)
+		return FS_COPY;
+	return 0;
```
I think this is supposed to return -1 for errors.
```suggestion
	return -1;
```

daltonbohning commented on src/utils/daos.c:
```diff
@@ -927,6 +1014,23 @@ call_dfuse_ioctl(char *path, struct dfuse_il_reply *reply)
 }
 
+static int
+fs_op_hdlr(struct cmd_args_s *ap)
```
We need to make sure both src and dst are provided.
For example:
`daos fs copy` gives "dfs_lookup / failed"
But really, we should make sure we have the required arguments first.

dsikich replied:
@daltonbohning I think we can save this for the next iteration. I will add a TODO since this I think deserves its own function to check all the arguments.

daltonbohning replied:
@dsikich Sure! There will also be a test case for this as well

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,758 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				/* Recursively call "fs_copy" with the new path. */
+				rc = fs_copy(daos_src_file, daos_dst_file, path, fs_dst_prefix);
+			}
+		}
```
@daltonbohning do you mean return ENOTSUP? if it is not a DIR or regular file in these situations?

daltonbohning replied:
@dsikich Yes. instead of silently skipping over anything that isn't a file or directory, I think we should return an error. Or, at the very least, log a message/error so there is a record of it

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,758 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	}
+	dfs_lookup(daos_file->dfs, dir, O_RDWR, &dirp->dir, NULL, NULL);
+	if (dirp->dir == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed", dir);
```
```suggestion
		fprintf(stderr, "dfs_lookup %s failed\n", dir);
```

daltonbohning replied:
Since we're using `fprintf` directly, I think we need newlines so the output doesn't get jumbled

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,758 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	int rc = dfs_read(daos_file->dfs, daos_file->obj, &sgl,
+			daos_file->offset, &got_size, NULL); 
+	if (rc) {
+		fprintf(stderr, "dfs_read %s failed (%d %s)",
```
```suggestion
		fprintf(stderr, "dfs_read %s failed (%d %s)\n",
```

daltonbohning reviewed (commented):
I think we need to include newlines in the error messages, since we're using `fprintf` directly, which doesn't add a newline by default.

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -2227,12 +2242,13 @@ fs_copy_connect(daos_file_t *daos_src_file,
 				DP_UUID(ap->dst_c_uuid));
 		}
 	}
+out:
```
@dsikich In a future iteration, we'll probably want to make sure we handle any necessary cleanup for any errors. For example, if we mount dfs but container creation fails, we need to make sure we unmount the container.

mchaarawi replied:
sorry, but this should be done in this PR, not a future iteration. we can't have daos leave containers created or cont / dfs handles dangling around. container handles do consumer resources on the server side.

dsikich replied:
@mchaarawi @daltonbohning added the cleanup for these cases

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1535,6 +1536,756 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+}
+
+static int
+daos_file_stat(daos_file_t *daos_file, const char *path, struct stat *buf)
```
Thanks!

mchaarawi commented on src/utils/daos_hdlr.h:
```diff
@@ -69,11 +92,22 @@ struct cmd_args_s {
+	uuid_t			dst_p_uuid;	/* --dst_pool */
+	uuid_t			src_c_uuid;	/* --src_cont */
+	uuid_t			dst_c_uuid;	/* --dst_cont */
+	char			*src_mdsrv_str;	/* --src_svc */
```
the svc is not required in master anymore. so we don't need to add it here.

mchaarawi commented on src/utils/daos.c:
```diff
@@ -560,9 +578,10 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
-	ap->p_op = -1;
-	ap->c_op = -1;
-	ap->o_op = -1;
+	ap->p_op  = -1;
```
i don't know why you added a space for the first free. maybe it's a tab and github isn't showing this properly.

mchaarawi commented on src/utils/daos.c:
```diff
@@ -630,6 +657,46 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 				D_GOTO(out_free, rc = RC_NO_HELP);
 			}
 			break;
+
```
i think we should consider here the feedback that was given before.
instead of needing to always specify --src-pool ... --src-cont ... --dst_pool, etc..
maybe have 1 string instead:
--src={daos://pool/container or path} --dst={daos://pool/container or path}

what do you think? (maybe consult with others here like Bruno and Johann)

dsikich replied:
@mchaarawi @johannlombardi @bfaccini I do agree having one string would simplify things. Is this something you want to handle in this PR? I ask because I imagine this is going to require adding some string parsing functions.

bfaccini replied:
well, I tend to agree with Mohamad, that would simplify a bit the command syntax and also have it look more like copy-commands common syntax. But this could also be part of a follow-on patch.

dsikich replied:
@mchaarawi @daltonbohning @bfaccini The only concern I have about making this change in this PR is that then the data mover interface that is part of mpifileutils (dcp) will be different. Is it better if we leave it as is until we can update the other data mover interfaces?

mchaarawi replied:
the problem if we don't make it is that people will start using this, and then we change it and they will have to change scripts / etc. similar to an API change, we should try to avoid unnecessary interface changes.

dsikich replied:
@mchaarawi ok I will work on getting this added here

mchaarawi commented on src/utils/daos.c:
```diff
@@ -630,6 +657,46 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
+				D_GOTO(out_free, rc = RC_NO_HELP);
+			}
+			break;
+		case 'X':
```
not needed anymore (latest master)

mchaarawi commented on src/utils/daos.c:
```diff
@@ -630,6 +657,46 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
+				D_GOTO(out_free, rc = RC_NO_HELP);
+			ap->src_mdsrv = daos_rank_list_parse(ap->src_mdsrv_str, ",");
+			break;
+		case 'x':
```
ditto

mchaarawi commented on src/utils/daos.c:
```diff
@@ -662,6 +729,16 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 			if (ap->path == NULL)
 				D_GOTO(out_free, rc = RC_NO_HELP);
 			break;
+		case 'R':
```
see comment on single src/dst options

mchaarawi commented on src/utils/daos.c:
```diff
@@ -927,6 +1014,23 @@ call_dfuse_ioctl(char *path, struct dfuse_il_reply *reply)
+fs_op_hdlr(struct cmd_args_s *ap)
+{
+	int rc = 0;
+	enum fs_op op;
```
checkpatch should have complained here. need an empty line after functiona variable declarations.

mchaarawi commented on src/utils/daos.c:
```diff
@@ -927,6 +1014,23 @@ call_dfuse_ioctl(char *path, struct dfuse_il_reply *reply)
+		rc = fs_copy_hdlr(ap);
+		break;
+	default:
+		break;
```
should be an assertion here to make sure.

mchaarawi commented on src/utils/daos.c:
```diff
@@ -1384,7 +1384,7 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 		FIRST_LEVEL_HELP();
-	} else if (strcmp(argv[2], "filesystem") != 0) {
+	} else if (strcmp(argv[2], "filesystem") != 0 || strcmp(argv[2], "fs") != 0) {
 		if (strcmp(argv[3], "copy") == 0) {
```
yes this was a very unfortunate bug in the daos tool that users heavily complained about

mchaarawi reviewed (changes requested):
sorry i have not had a chance yet to review the core part of the code in daos_hdlr.c
please make sure you follow the PR guidelines in:
https://wiki.hpdd.intel.com/display/DC/Commit+Comments
(you don't have a component in the title, which in this case should be daos since this is the daos tool).

dsikich commented:
@mchaarawi @daltonbohning thank you for taking the time to review and give feedback. I will submit the suggested changes later this afternoon. I have an appt at 3pm today so I will have to do it when I get back. I know lots of people will be on vacation next week, so I can maybe add some more people as reviewers since I know @daltonbohning will be out, or we can review my edits the following week.

dsikich commented:
> @mchaarawi @daltonbohning thank you for taking the time to review and give feedback. I will submit the suggested changes later this afternoon. I have an appt at 3pm today so I will have to do it when I get back. I know lots of people will be on vacation next week, so I can maybe add some more people as reviewers since I know @daltonbohning will be out, or we can review my edits the following week.

@daltonbohning @mchaarawi submitted update for week after next for review

daltonbohning commented on src/utils/daos.c:
```diff
@@ -1280,6 +1382,19 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 
 	if (argc <= 2) {
 		FIRST_LEVEL_HELP();
```
In the definition for `FIRST_LEVEL_HELP` (above on line ~1300), we should have an entry for filesystem (fs).
This will add an entry for `daos help`

dsikich replied:
@daltonbohning added this in latest commit

daltonbohning reviewed (commented):
Running some test cases:

**_When no source or dest supplied_**
```
daos fs copy
...
dfs_lookup / failed
Source is not a directory: /
daos: src/utils/daos.c:1011: fs_op_hdlr: Assertion `rc == 0' failed.
Aborted
```

_**For any copy**_
```
daos fs copy --dst-pool $pool1 --dst-cont $p1cont1 source /dest
...
unexpected extra argument : 'source'
error parsing command line arguments
```
I think this is because of:
https://github.com/daos-stack/daos/blob/59b0b7d1864fcc5c2846b5e2730ed64325a2c784/src/utils/daos.c#L843-L848
Maybe for "filesystem" we can alter the if statement to allow two more arguments (source and destination)

dsikich commented:
> Running some test cases:
> 
> **_When no source or dest supplied_**
> 
> ```
> daos fs copy
> ...
> dfs_lookup / failed
> Source is not a directory: /
> daos: src/utils/daos.c:1011: fs_op_hdlr: Assertion `rc == 0' failed.
> Aborted
> ```
> 
> _**For any copy**_
> 
> ```
> daos fs copy --dst-pool $pool1 --dst-cont $p1cont1 source /dest
> ...
> unexpected extra argument : 'source'
> error parsing command line arguments
> ```
> 
> I think this is because of:
> 
> https://github.com/daos-stack/daos/blob/59b0b7d1864fcc5c2846b5e2730ed64325a2c784/src/utils/daos.c#L843-L848
> 
> 
> Maybe for "filesystem" we can alter the if statement to allow two more arguments (source and destination)

On the "for any copy" test I didn't intend for it to be used that way. I explicitly added --src-path and --dst-path options. I suppose doing it the way you tested it is more consistent with how dcp works.

daltonbohning commented:
> On the "for any copy" test I didn't intend for it to be used that way. I explicitly added --src-path and --dst-path options. I suppose doing it the way you tested it is more consistent with how dcp works.

Oh, I overlooked that. I was trying to copy-paste some test from dcp to this utility. I think maybe we should stick with --src-path and --dst-path, since that is how the daos utility seems to do things. But more generally, I think once we decide on the simplified format - e.g. daos/pool/containter/path - this won't be an issue.

mchaarawi commented on Jenkinsfile:
```diff
@@ -322,6 +322,7 @@ boolean skip_scan_rpms_centos7() {
 
 boolean skip_ftest_hw(String size) {
     return env.DAOS_STACK_CI_HARDWARE_SKIP == 'true' ||
+           skip_stage('func-test') ||
```
im not sure what this is about?

daltonbohning replied:
This is a change that just landed in master today. I wonder if somehow git got confused?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		size = -1;
+	}
+
+	daos_file->offset += (daos_off_t)size;
```
this is not threadsafe

mchaarawi replied:
also in case of error, this would corrupt the offset since you set size to -1

dsikich replied:
@mchaarawi I thought the fs_copy was only for serial copies? The parallel version will be in dcp.

mchaarawi replied:
fair enough. my point was only if you are moving this function into a "common" library, this would require to be threadsafe. not really about parallelism in an MPI manner, but just multiple threads (single address space).

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+}
+
+static ssize_t
+daos_write(struct daos_file_t *daos_file, char *file, void *buf, size_t size)
```
daos_ prefix is for public API functions, so for internal functions we shouldn't be using it. (applies to all other static functions here)

daltonbohning replied:
I think this ties into [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098), where we need to decide where we want these functions to live, and how to name them.

daltonbohning replied:
But I'm not sure what the name should be in the meantime.

mchaarawi replied:
yea sure. if we are adding those to libdfs, it would be a dfs_ prefix. or maybe dfs_hlpr_ ..
but in this PR, since this is static function in this file only, it doesn't need a prefix. maybe just write_dfs_*

dsikich replied:
@mchaarawi @daltonbohning added a _dfs to the end of the I/O functions and the file struct, instead of using the daos_ prefix

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	d_iov_set(&iov, buf, size);
+
+	/* define scatter-gather list for dfs_write */
+	d_sg_list_t sgl;
```
for code formatting, variable declarations are always to be done first in the function.
there is probably some test code that does this, but we should not be doing that.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	sgl.sg_nr_out = 1;
+
+	rc = dfs_write(daos_file->dfs, daos_file->obj, &sgl, daos_file->offset, NULL); 
+
```
nitpick: extra line

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		fprintf(stderr, "dfs_write %s failed (%d %s)\n",
+			file, rc, strerror(rc));
+		errno = rc;
+		size = -1;
```
size is unsigned?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+}
+
+static ssize_t
+daos_file_write(daos_file_t *daos_file, char* file,
```
please check the format of arguments:
char \*file vs char\* file
many other occurrences throughout

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	rc = dfs_lookup(daos_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+	}
```
in case of error, why do we still continue to the open?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	return num_bytes_written;
+}
+
+static void 
```
this should return an error, why void?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		fprintf(stderr, "dfs_open %s failed (%d)\n", name, rc);
+}
+
+void
```
no error handling here?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	dfs_obj_t *parent = NULL;
+	char *name = NULL;
+	char *dname = NULL;
+	parse_filename(path, &name, &dname);
```
checkpatch should complain really here.. need an empty line between declarations and the function body

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	char *dname = NULL;
+	parse_filename(path, &name, &dname);
+    
+	assert(dname);
```
why assert not error? is it guaranteed that user can't trigger this assertion

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	parse_filename(path, &name, &dname);
+    
+	assert(dname);
+	if (name && strcmp(name, "/") != 0) {
```
i don't understand this if condition.
how can name be "/" in this case?
and if name is NULL, why do we not return an error??

daltonbohning replied:
I think this is carry-over from mpifileutils trying to create the directory "/" in DFS.
Maybe we don't need that here, since this doesn't work exactly the same way as mpifileutils.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	if (daos_file->type == POSIX) {
+		rc = mkdir(dir, mode);
+	} else if (daos_file->type == DAOS) {
+		rc = daos_mkdir(daos_file, dir);
```
you have the mode here from the input parameters, why are you not passing it to DFS?

dsikich replied:
@mchaarawi updated, also added the chmod functions to update permissions to match the source file after copying

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	} else if (daos_file->type == DAOS) {
+		rc = daos_mkdir(daos_file, dir);
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n", dir, daos_file->type);
```
rc is not set to error invalid, so function is returning success in this case..

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	dfs_lookup(daos_file->dfs, dir, O_RDWR, &dirp->dir, NULL, NULL);
+	if (dirp->dir == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed\n", dir);
+		errno = ENOENT;
```
dfs_lookup returns a positive errno, so why set this to ENOENT and not what dfs_lookup returns?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		}
+		goto ret;
+	}
+	assert(daos_anchor_is_eof(&dirp->anchor));
```
this assertion is redundant

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		if (rc) {
+			fprintf(stderr, "dfs_readdir failed (%d %s)\n", rc, strerror(rc));
+			dirp->num_ents = 0;
+			errno = ENOENT;
```
readdir also returns a proper errno, so enoent is wrong here

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	rc = dfs_lookup(daos_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+	if (parent == NULL) {
+		fprintf(stderr, "dfs_lookup %s failed \n", dir_name);
+		errno = ENOENT;
```
same

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+
+static ssize_t
+daos_read(daos_file_t *daos_file,
+	const char* file,
```
fromatting

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	} else {
+		fprintf(stderr, "File type not known: %s type=%d\n", file, daos_file->type);
+	}
+	return got_size;
```
in case of error you should return -1?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	} else if (daos_file->type == DAOS) {
+		rc = daos_closedir(dirp);
+	} else {
+		fprintf(stderr, "File type not known, type=%d\n", daos_file->type);
```
error code

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		if (rc == 0)
+			daos_file->obj = NULL;
+	} else {
+		fprintf(stderr, "File type not known, file=%s, type=%d\n",
```
error code

mchaarawi reviewed (changes requested):
i got the to fs_copy() function but ran out of time today to finish

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+			fprintf(stderr, "failed to connect to destination pool: %d\n", rc);
+			/* if connection to source fails, do not continue */
+			D_GOTO(out, rc);
+		} else {
```
please remove the else statement

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				daos_src_file->type = POSIX;
+			}
+		}
+	}
```
there should be an error returned here in a final else, why continue if this failed?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				}
+			}
+		}
+	} else {
```
} else if (ap->src_path != NULL) {

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				ap->type = (*src_dattr).da_type;
+				uuid_copy(ap->src_p_uuid, (*src_dattr).da_puuid);
+				uuid_copy(ap->src_c_uuid, (*src_dattr).da_cuuid);
+				rc = daos_pool_connect(ap->src_p_uuid, ap->sysname, NULL,
```
seems like a duplicate of the connect / open/ mount above.
to avoid duplication we can do the checks and resolve the pool and cont uuid from path or input paramters then do a single point where we do connect / open /mount

dsikich replied:
@mchaarawi added

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				ap->type = (*dst_dattr).da_type;
+				uuid_copy(ap->dst_p_uuid, (*dst_dattr).da_puuid);
+				uuid_copy(ap->dst_c_uuid, (*dst_dattr).da_cuuid);
+				rc = daos_pool_connect(ap->dst_p_uuid, ap->sysname, NULL,
```
same comment on code duplication

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				daos_dst_file->type = POSIX;
+			}
+		}
+	}
```
can fail here if we fail the destination connect

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+				D_GOTO(out, rc);
+			}
+		}
+	} else {
```
else if on one line please to avoid extra tab

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+						ap->dst_path, rc);
+					D_GOTO(out, rc);
+				}
+			} else {
```
so if path is not NULL, but not a duns path we create it? where is the call to duns_create()?
if it's a valid path but not a UNS one, we should fail though.
the only way we should create the duns path is if it doesn't  exist. the duns_resolve returns ENOENT

dsikich replied:
@mchaarawi not sure I follow what you mean here, if if the path is not null, and it is is not a duns path, then I assume/set it to a POSIX type

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	}
+	/* create DAOS POSIX destination container if it doesn't exist */
+	if (daos_dst_file->type == DAOS) {
+		if (daos_uuid_valid(ap->src_c_uuid) && !daos_uuid_valid(ap->dst_c_uuid)) {
```
i guess i don't get this.
if the user doesn't pass a dst uuid why do we even try to do the cont open above?
and what if the user passes a rubbish uuid? we should fail in this case not generate one and create IMO.

dsikich replied:
@mchaarawi @daltonbohning I removed the case where a dest container is created if one isn't given, maybe we can support something like this but I think it sounds like that case needs to be worked out better first. Is creating a dest container if it doesn't exist something we eventually want to support for this tool?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	if (daos_dst_file->type == DAOS) {
+		if (daos_uuid_valid(ap->src_c_uuid) && !daos_uuid_valid(ap->dst_c_uuid)) {
+			/* create POSIX destination container if it doesn't exist */
+	    		if (uuid_is_null(ap->c_uuid))
```
ap->dst_c_uuid?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+}
+
+static void
+daos_file_set_defaults(daos_file_t *daos_file)
```
this can be an inline function

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	if (daos_src_file.type == POSIX && daos_dst_file.type == DAOS) {
+		ap->dst_path = "/";
+		parse_filename(ap->src_path, &name, &dname);
+		dfs_set_prefix(daos_dst_file.dfs, dname);
```
could you please explain why we do this here?
the dname is on the source path which is POSIX, but we set it on the dst dfs mount?

dsikich replied:
@mchaarawi yeah true setting it here isn't necessary since it already gets set

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		ap->src_path = "/";
+		ap->dst_path = "/";
+		rc = fs_copy(&daos_src_file, &daos_dst_file,
+				ap->src_path, dname);
```
dname is NULL here?

dsikich replied:
@mchaarawi yeah, it won't be used in the DAOS -> DAOS case in fs_copy it is only used when the dest paths need to be constructed for POSIX

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+			D_GOTO(out, rc);
+	}
+
+	/* umount dfs, close conts, and disconnect pools */
```
the disconnect should be done also on error cases after you finish the connect

dsikich replied:
@mchaarawi added

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		int path_length = 0;
+		path_length = snprintf(filename, MAX_FILENAME, "%s/%s", dir_name, d_name);
+		if (path_length >= MAX_FILENAME) 
+			fprintf(stderr, "Path length is too long.\n");
```
just print errors and continue?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+			path_length = snprintf(dst_filename, MAX_FILENAME, "%s/%s", fs_dst_prefix, filename);
+			if (path_length >= MAX_FILENAME) 
+				fprintf(stderr, "Path length is too long.\n");
```
ditto

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		struct stat st;
+		rc = daos_file_lstat(daos_src_file, filename, &st);
+		if (rc)
+			fprintf(stderr, "Cannot stat path %s, %s\n",
```
ditto

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1561,6 +1562,817 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+
+		if (S_ISREG(st.st_mode)) {
+			if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+				daos_file_open(daos_src_file, filename, O_CREAT | O_RDWR, st.st_mode);
```
error handling..

dsikich commented:
@mchaarawi Thank you for the feedback. I believe I corrected and/or addressed all of your comments in my latest commit. I also added the interface to support --src={daos://pool/cont or path} and posix would be --dst={posix://a/b/c}

daltonbohning commented:
1. Maybe we need more error handling in case the source and destination doesn't match the correct format.
```
> daos fs copy --src a --dst b
...
failed to connect to destination pool: -1003
```
```
> daos fs copy --src {posix:./s} --dst {fake:./d}
...
failed to connect to destination pool: -1003
```

2. Do we intend for the `{` and `}` to be literal?
```
> daos fs copy --src {posix:./s} --dst {daos://$pool1/$p1cont1/}
// good

> daos fs copy --src posix:./s --dst daos://$pool1/$p1cont1/
// fail
```

3. These three seem to all work, which is maybe fine:
```
--dst {daos://$pool1/$p1cont1/} # //
--dst {daos:/$pool1/$p1cont1/}  # /
--dst {daos:$pool1/$p1cont1/}   # without /, but still at root of container
```

4. The destination directory doesn't seem to be named as given.
```
> daos fs copy --src {posix:./s} --dst {daos://$pool1/$p1cont1/d}
> ls -l dfuse
drwx------ 1 dbohninx dbohninx 64 Dec  7 17:48 s
// Has the same name as the source, but "d" was given for the destination
```
```
> daos fs copy --src {posix:./s/} --dst {daos:/$pool1/$p1cont1/many/dirs/deep/}
// The destination is still just "s" in the container
```

5. POSIX destination doesn't seem to work
```
// With relative path
> daos fs copy --src {daos://$pool1/$p1cont1/s/} --dst {posix:./d}
// no errors
> ls -l d
No such file or directory
```
```
// With absolute path
> daos fs copy --src {daos://$pool1/$p1cont1/s/} --dst {posix://tmp/dbohninx/test/d/}
// no errors
> ls -l d
No such file or directory
```

dsikich commented:
> 1. Maybe we need more error handling in case the source and destination doesn't match the correct format.
> 
> ```
> > daos fs copy --src a --dst b
> ...
> failed to connect to destination pool: -1003
> ```
> 
> ```
> > daos fs copy --src {posix:./s} --dst {fake:./d}
> ...
> failed to connect to destination pool: -1003
> ```
> 
> 1. Do we intend for the `{` and `}` to be literal?
> 
> ```
> > daos fs copy --src {posix:./s} --dst {daos://$pool1/$p1cont1/}
> // good
> 
> > daos fs copy --src posix:./s --dst daos://$pool1/$p1cont1/
> // fail
> ```
> 
> 1. These three seem to all work, which is maybe fine:
> 
> ```
> --dst {daos://$pool1/$p1cont1/} # //
> --dst {daos:/$pool1/$p1cont1/}  # /
> --dst {daos:$pool1/$p1cont1/}   # without /, but still at root of container
> ```
> 
> 1. The destination directory doesn't seem to be named as given.
> 
> ```
> > daos fs copy --src {posix:./s} --dst {daos://$pool1/$p1cont1/d}
> > ls -l dfuse
> drwx------ 1 dbohninx dbohninx 64 Dec  7 17:48 s
> // Has the same name as the source, but "d" was given for the destination
> ```
> 
> ```
> > daos fs copy --src {posix:./s/} --dst {daos:/$pool1/$p1cont1/many/dirs/deep/}
> // The destination is still just "s" in the container
> ```
> 
> 1. POSIX destination doesn't seem to work
> 
> ```
> // With relative path
> > daos fs copy --src {daos://$pool1/$p1cont1/s/} --dst {posix:./d}
> // no errors
> > ls -l d
> No such file or directory
> ```
> 
> ```
> // With absolute path
> > daos fs copy --src {daos://$pool1/$p1cont1/s/} --dst {posix://tmp/dbohninx/test/d/}
> // no errors
> > ls -l d
> No such file or directory
> ```

@daltonbohning 
it doesn't look like it is resolving relative paths

2. I don't know honestly, the braces probably aren't necessary 

4. It was intended to support only pool/cont or path, not both

5. My POSIX test cases were passing, but maybe in my parsing I didn't account for the extra / you have at the end, I can take a look

daltonbohning commented:
@dsikich 
> it doesn't look like it is resolving relative paths

Are you referring to this test case?
`daos fs copy --src a --dst b`
If so, I don't think it's a matter of relative paths, but rather that the format doesn't match "{posix://path}", etc.

> 2. I don't know honestly, the braces probably aren't necessary

@mchaarawi Did you mean for the braces to be literal?

> 3. It was intended to support only pool/cont or path, not both

I think it's probably fine to just ignore the extra `/` (as it is working now).

> 4. My POSIX test cases were passing, but maybe in my parsing I didn't account for the extra / you have at the end, I can take a look

I see what's going on now. This is just the same thing as the destination directory not being named as expected. I had the destination "s" in my working directory, which was just silently being overridden.

dsikich commented:
> @dsikich
> 
> > it doesn't look like it is resolving relative paths
> 
> Are you referring to this test case?
> `daos fs copy --src a --dst b`
> If so, I don't think it's a matter of relative paths, but rather that the format doesn't match "{posix://path}", etc.
> 
> > 1. I don't know honestly, the braces probably aren't necessary
> 
> @mchaarawi Did you mean for the braces to be literal?
> 
> > 1. It was intended to support only pool/cont or path, not both
> 
> I think it's probably fine to just ignore the extra `/` (as it is working now).
> 
> > 1. My POSIX test cases were passing, but maybe in my parsing I didn't account for the extra / you have at the end, I can take a look
> 
> I see what's going on now. This is just the same thing as the destination directory not being named as expected. I had the destination "s" in my working directory, which was just silently being overridden.

@daltonbohning thanks, I am going to rebase my cont_copy branch since the OID table is merged now. Then I will work on these, and I can remove the braces then if need be, shouldn't be a big deal to do that.

daltonbohning commented:
6. Source path seems to be ignored
```
> daos fs copy --src {daos:$pool1/$p1cont1/fake/path} --dst {posix:/tmp/dbohninx/test/d/}
// Uses the container root / as the source
```

daltonbohning commented:
> @daltonbohning thanks, I am going to rebase my cont_copy branch since the OID table is merged now. Then I will work on these, and I can remove the braces then if need be, shouldn't be a big deal to do that.

Sure thanks! I'm not sure about the format either, so I guess we can keep that until told otherwise.

daltonbohning commented:
@dsikich I think the help options also need to be updated to show `--src={...}` and `--dst={...}` instead of `--src-pool`, `--src-cont`, etc.

dsikich commented:
@daltonbohning updated the help menu, and added some checking for if the file type is not supported on the new interface. Also, I removed the braces. I ran all of your test cases and they all passed for me, so here's to hoping it is the same for you. I have not yet added support for parsing container subsets in DFS for the new interface, because I want to make sure this is relatively stable first.

daltonbohning commented:
@dsikich
Do we want --src and --dst to be on separate lines?
```
> daos help fs
filesystem copy options (copy):
        --src={daos | posix}:{pool/cont | path} --dst={daos | posix}:{pool/cont | path}
```

Also, `daos help fs` is different from `daos help fs copy`
```
> daos help fs copy
filesystem copy options (copy):
        --src-pool=UUID    src pool UUID
        --dst-pool=UUID    dst pool UUID
        --src-cont=UUID    src cont UUID
        --dst-cont=UUID    dst cont UUID
        --src-path=PATH    POSIX or DFS source path
        --dst-path=PATH    POSIX or DFS destination path
```

I'm getting a segmentation fault:
```
> echo $pool1
1d4f69ab-d9ce-4264-864f-5a1d5d60b227
> echo $p1cont1
8e90314f-4aa9-4ecc-ac20-c6b053cb42d4

> valgrind daos fs copy --src daos:$pool1:$p1cont1 --dst posix:/tmp/dbohninx/dest
Access not within mapped region at address 0x0
==44792==    at 0x4C2D0F2: strlen (vg_replace_strmem.c:461)
==44792==    by 0x53DC468: uuid_parse (in /usr/lib64/libuuid.so.1.3.0)
==44792==    by 0x40A508: fs_copy_parse_uuids.isra.4 (daos_hdlr.c:2458)
==44792==    by 0x411ECB: fs_copy_parse (daos_hdlr.c:2505)
==44792==    by 0x411ECB: fs_copy_hdlr (daos_hdlr.c:2563)
==44792==    by 0x4099A8: fs_op_hdlr (daos.c:1016)
==44792==    by 0x404AD2: main (daos.c:1613)
```

dsikich commented:
> @dsikich
> Do we want --src and --dst to be on separate lines?
> 
> ```
> > daos help fs
> filesystem copy options (copy):
>         --src={daos | posix}:{pool/cont | path} --dst={daos | posix}:{pool/cont | path}
> ```
> 
> Also, `daos help fs` is different from `daos help fs copy`
> 
> ```
> > daos help fs copy
> filesystem copy options (copy):
>         --src-pool=UUID    src pool UUID
>         --dst-pool=UUID    dst pool UUID
>         --src-cont=UUID    src cont UUID
>         --dst-cont=UUID    dst cont UUID
>         --src-path=PATH    POSIX or DFS source path
>         --dst-path=PATH    POSIX or DFS destination path
> ```
> 
> I'm getting a segmentation fault:
> 
> ```
> > echo $pool1
> 1d4f69ab-d9ce-4264-864f-5a1d5d60b227
> > echo $p1cont1
> 8e90314f-4aa9-4ecc-ac20-c6b053cb42d4
> 
> > valgrind daos fs copy --src daos:$pool1:$p1cont1 --dst posix:/tmp/dbohninx/dest
> Access not within mapped region at address 0x0
> ==44792==    at 0x4C2D0F2: strlen (vg_replace_strmem.c:461)
> ==44792==    by 0x53DC468: uuid_parse (in /usr/lib64/libuuid.so.1.3.0)
> ==44792==    by 0x40A508: fs_copy_parse_uuids.isra.4 (daos_hdlr.c:2458)
> ==44792==    by 0x411ECB: fs_copy_parse (daos_hdlr.c:2505)
> ==44792==    by 0x411ECB: fs_copy_hdlr (daos_hdlr.c:2563)
> ==44792==    by 0x4099A8: fs_op_hdlr (daos.c:1016)
> ==44792==    by 0x404AD2: main (daos.c:1613)
> ```

@daltonbohning did you mean to put an error in the fs copy you did with valgrind? There should not be a ":" between pool1 and p1cont1. There should be a "/" there instead

dsikich commented:
@daltonbohning do you think its easier to have a ":" instead of a "/" to separate pools/conts?

daltonbohning commented:
> did you mean to put an error in the fs copy you did with valgrind? There should not be a ":" between pool1 and p1cont1. There should be a "/" there instead

@dsikich Ah, that was a typo. But I guess maybe we should account for this too?

daltonbohning commented:
> @daltonbohning do you think its easier to have a ":" instead of a "/" to separate pools/conts?

I think the "/" is better, since that is orthogonal to how dfuse allows "dfuse_mount/$pool/$cont/path/in/cont".
But I do think we should have error checking for this. 
I think after `pool = strtok_r(str, "/", str_ptr);` we just need to make sure `pool` is not null

daltonbohning commented:
@dsikich Another note about the help options.
Using this as a reference:
```
> daos help cont create
...
--properties=<name>:<value>[,<name>:<value>,...]
                           supported prop names are label, cksum,
                                cksum_size, srv_cksum, dedup
                                dedup_th, compression, encryption
                           label value can be any string
...
```

I wonder if we want to do something like:
```
> daos help fs copy
--src=<type>:<pool/cont | path>
                           supported types are daos, posix

--dst=<type>:<pool/cont | path>
```

daltonbohning commented:
@dsikich Minor Consideration: `dcp` handles trailing slashes differently than paths without trailing slashes.
For example, `dcp` essentially interprets "path/" as "path/*".
This is handled by the copy_into_dir logic, etc., and also depends on whether the destination already exists.
For what it's worth, linux `cp` does NOT seem to do this.

I guess this is something we might want to get other feedback on, but here is an example:
Running dcp for POSIX -> DAOS
```
> dcp --daos-dst-pool $pool1 --daos-dst-cont $p1cont1 source/ /
> ls -R dfuse
dfuse:
file1
```

Running daos fs copy for POSIX -> DAOS
```
> daos fs copy --src posix:./source/ --dst daos:$pool1/$p1cont1/
> ls -R dfuse
dfuse:
source

dfuse/source:
file1
```

dsikich commented:
@daltonbohning alright added ability to copy cont subsets, fixed help options, and added the null check on the pool and conts

dsikich commented:
@mchaarawi @daltonbohning My goal is to at least get this PR approved this week before the holidays? @mchaarawi if you don't have time do you suggest anyone else that could review this PR?

mchaarawi commented:
> @mchaarawi @daltonbohning My goal is to at least get this PR approved this week before the holidays? @mchaarawi if you don't have time do you suggest anyone else that could review this PR?

i will review it again. it wasn't showing up in my list for items to review until now

mchaarawi reviewed (commented):
are you still updating? you asked for a review then the file just refreshed as i was reviewing..

daltonbohning commented on src/utils/daos.c:
```diff
@@ -45,15 +45,25 @@
 #include <daos/rpc.h>
 #include <daos/debug.h>
 #include <daos/object.h>
+#include <sys/stat.h>
```
```suggestion
```
This is included above on line 37. I think we can remove this one.

dsikich commented:
> are you still updating? you asked for a review then the file just refreshed as i was reviewing..

@mchaarawi sorry about that, @daltonbohning asked me to merge it into the latest master then I noticed I forgot to delete the old options (src-pool, dst-pool, etc.) so all I did was delete those. I'm not updating anymore.

daltonbohning commented on src/utils/daos_hdlr.h:
```diff
@@ -62,6 +66,25 @@ enum obj_op {
+#define MAX_FILENAME 256
+
+struct dfs_daos_t {
+	dfs_obj_t* dir;
```
I'm getting this compile error: (master seems to compile without issue)
```
In file included from src/utils/daos_autotest.c:38:0:
src/utils/daos_hdlr.h:73:2: error: unknown type name ‘dfs_obj_t’
  dfs_obj_t* dir;
```

dsikich replied:
> I'm getting this compile error: (master seems to compile without issue)
> 
> ```
> In file included from src/utils/daos_autotest.c:38:0:
> src/utils/daos_hdlr.h:73:2: error: unknown type name ‘dfs_obj_t’
>   dfs_obj_t* dir;
> ```

@daltonbohning honestly not sure what happened there. I was just building this about a week ago, that type is defined in daos_fs.h, which has been added. I'm looking into it now.

dsikich replied:
@daltonbohning are you adding daos_fs.h?

daltonbohning replied:
It seems that `daos_hdlr.h` is now included from `daos_autotest.c`. And `daos_hdlr.h` doesn't include `daos_fs.h`, so the types aren't defined.
I wonder if we can actually just put the structs from `daos_hdlr.h` into `daos_hdlr.c`, since they are only being used within `daos_hdlr.c` anyway?

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,6 +1567,1119 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+			&ap->cont, src_cont_info, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to open source container: %d\n", rc);
+			rc = daos_pool_disconnect(ap->pool, NULL);
```
minor, style:
this is technically correct, but you can do this easier with just gotos. something like:
https://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c#L1299
instead of repearing the cascading close/disconnect on every phase.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,6 +1567,1119 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	}
+
+	/* open dst pool, dst cont, and mount dfs */
+	if (dst_file_dfs->type == DAOS) {
```
so if source open/connect failed, why do the destination?

dsikich replied:
@mchaarawi if there is a failure connecting/opening src then it doesn't get to this point the function should return

mchaarawi commented on src/utils/daos_hdlr.h:
```diff
@@ -62,6 +66,25 @@ enum obj_op {
+	daos_off_t offset;
+	dfs_obj_t* obj;
+	dfs_t* dfs;
+} file_dfs_t;
```
do we need the above 2 structs in this header? can we just have them local to the file you are using them?

daltonbohning replied:
Agreed. This is also causing a compile issue (noted previously)

dsikich replied:
@mchaarawi yeah I will move that.

dsikich commented on src/utils/daos_hdlr.h:
```diff
@@ -70,11 +93,18 @@ struct cmd_args_s {
+	enum fs_op		fs_op;		/* filesystem sub-command */
 	char			*sysname;	/* --sys-name or --sys */
 	uuid_t			p_uuid;		/* --pool */
+	uuid_t			src_p_uuid;	/* --src_pool */
```
@mchaarawi well I suppose I fill them in after parsing the uuids. I did that because some of the other options have the uuid_t variable stored in the args struct, but I don't think it really has to be stored there.

dsikich commented on src/utils/daos.c:
```diff
@@ -816,7 +875,9 @@ common_op_parse_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 	 * and client will query management service for rank list.
 	 */
-	ARGS_VERIFY_MDSRV(ap, out_free, rc = RC_PRINT_HELP);
+        if (ap->fs_op != FS_COPY) {
```
@mchaarawi I think it can be removed since svc is no longer required to be passed in, but I was doing that because I had separate src and dst svcls and that was only checking one svc

dsikich commented on src/utils/daos.c:
```diff
@@ -1223,13 +1312,25 @@ do { \
+	" filesystem copy options (copy):\n" \
+	"	--src=<type>:<pool/cont | path>\n" \
+	"	--dst=<type>:<pool/cont | path>\n" \
+	"	\t supported types are daos, posix\n"); \
```
@mchaarawi ok, yeah I see what you mean.

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,6 +1567,1119 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		size = -1;
+	}
+
+	file_dfs->offset += (daos_off_t)size;
```
updated. Must have missed it mixed in with all of the other comments :-/

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,6 +1567,1119 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	}
+	dirp->num_ents = NUM_DIRENTS;
+	int rc;
+	while (!daos_anchor_is_eof(&dirp->anchor)) {
```
@mchaarawi in fs_copy now

dsikich commented:
@daltonbohning @mchaarawi I added a commit to address your feedback on your reviews

daltonbohning commented on src/utils/daos.c:
```diff
@@ -1284,6 +1349,10 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 
 	if (argc <= 2) {
 		FIRST_LEVEL_HELP();
+	} else if (strcmp(argv[2], "filesystem") == 0 || strcmp(argv[2], "fs") == 0) {
+		if (strcmp(argv[3], "copy") == 0) {
```
I think we still need to make sure argv[3] is valid here.
Test cases:
```
> daos help fs
daos command (v1.1.2.1)
Segmentation fault

> daos help fs copy
# works as expected
```

daltonbohning commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	p_dst = strtok_r(NULL, ":", &dst_saveptr);
+
+	/* check for src DAOS pool/cont or UNS path */ 
+	if (strcmp(src_type, "daos") == 0) {
```
At one point @mchaarawi recommended something like:
```
daos://<pool>/<cont>/<path>
```
This parsing doesn't seem too support the preceding `//`. Instead, it expects:
```
daos:<pool>/<cont>/<path>
```
For example:
```
> daos fs copy --src daos://$pool1/$p1cont1/source --dst dest
dfs_lookup /tmp/dbohninx/03 failed, 2
```
I'm okay with either way, but I'm not sure what is _**actually**_ expected.
If we want to enforce `daos://` exactly, maybe we can do something like this:
```
// not tested
if (strncmp(src, "daos://", 8) == 0) {
    src_type = "daos"; // Treat as daos
    src += 8; // Just move pointer to where the pool uuid should be
}
```
Then, anything else would just be treated as POSIX, including incorrect variants like these: `daos:/<pool>`, `daos:<pool>`, `daos<pool>`
In which case the POSX path probably doesn't exist, which would be caught by the other error handling.

daltonbohning replied:
Also, some other error checking for this:
```
> daos fs copy --src daos:///c3501a11-697a-4bb8-a8b6-ef45cccd911a/ --dst dst
...
segmentation fault
```

dsikich replied:
@daltonbohning I can add this if this is what we want to enforce.

daltonbohning replied:
I'll let @mchaarawi comment here.

dsikich replied:
@daltonbohning the more I think about this I think enforcing the daos:// is better because it prevents these issues/errors you are seeing better

dsikich replied:
@daltonbohning updating this on my latest commit to enforce daos:// prefix when pool/cont are specified

daltonbohning reviewed (commented):
@dsikich Mostly minor style and error checking, except for the --src and --dst format, which I would like @mchaarawi's input on.
Specifically:
`--src daos:<pool>/<cont>/<path>` // Should this work?
`--src daos:/<pool>/<cont>/<path>` // Should this work?
`--src daos://<pool>/<cont>/<path>` // Should this work?

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	return rc;
+}
+
+static int fs_copy_parse_uuids(char* str,
```
@daltonbohning dang I really thought I got all of these.

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	}
+out:
+	if (parent != NULL) {
+		rc = dfs_release(parent);
```
@daltonbohning well, not entirely sure the best way to do this, because if I just return if dfs_mkdir fails, then I don't release the parent or memory from dirname, name? I think that when dfs_mkdir runs it does log a message, so will dfs_release. I guess the function will return though with the error code from dfs_release if it fails.

daltonbohning replied:
@dsikich Right, we do still need to call dfs_release if dfs_mkdir fails. But if dfs_mkdir fails, and dfs_release succeeds, then rc=0 will be returned.
In mpifileutils, I recently added something like:
```
tmp_rc = dfs_release(...)
if (tmp_rc && (rc != -1)) {
    // log error and set errno accordingly
}
```
Basically, ignore a failed dfs_release if and only if dfs_mkdir already failed.
In dfs.c, I think this is handled in some places by just simply ignoring the rc from dfs_release in situations like this. For functions like "close", etc., a failed release is definitely problem. But for "transient" cases like this, it seems that dfs.c doesn't care.
But... I'm not sure if that's intentional or not.
Maybe @mchaarawi has an opinion.

dsikich replied:
@daltonbohning using tmp_rc for now

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+		fprintf(stderr, "dfs_release failed (%d %s)\n",
+			rc, strerror(rc));
+		errno = rc;
+		rc = 1;
```
@daltonbohning yeah, since it was mentioned I was returning both negative and positive error codes I went with positive. Since the dfs_ functions all return positive error codes. What I thought I would do is check if there was an error on a POSIX function, then set the rc to the errno since the POSIX functions do set that, and its more consistent with what the dfs_ functions return?

daltonbohning replied:
@dsikich Returning positive errno codes is more consistent with the DFS API, but not with the POSIX API.
Most of the POSIX syscalls return -1 on error and set errno accordingly.
So if we want these DFS wrapper functions to be consistent with the POSIX equivalents, I think we should return -1 on error and set errno accordingly.
Thoughts?

mchaarawi replied:
for wrappers i prefer to match what DFS does for the error code. it's usually tricky to mess with errno especially at a user space level. but i don't have a strong opinion here. i remember once before Ashley did, so maybe discuss with him?

daltonbohning replied:
@mchaarawi I was thinking maybe we could cross that bridge as part of [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098). I.e. as long as error checks are in place here, we can worry about setting errno vs returning an errno later, which will be discussed as part of [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098).
Are you okay with this?

mchaarawi replied:
yes but in this case then we should either go back to return -1 or the actual error code.
1 is weird to me and wrong in the case of read.

dsikich replied:
@mchaarawi yeah returning 1 for bytes_read is weird I can update that.

daltonbohning replied:
Yes, I agree

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -1566,11 +1598,1197 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+{
+	int rc = dfs_release(file_dfs->obj);
+
+	if (rc) {
+		fprintf(stderr, "dfs_close %s failed (%d %s)\n",
+			file, rc, strerror(rc));
+		errno = rc;
+	}
```
@daltonbohning well I actually want to return the error code from these functions now, so I just updated to return that without setting 1/-1, so it should return whatever errno was set to. The POSIX functions set -1 on error, and then set errno, so not sure the best way to be consistent when the dfs_ functions which return errno code on failure. I've just been updating to return errno code always on failure for POSIX in the file_ wrappers.

daltonbohning replied:
@dsikich Yeah, I'm also not sure which is the best way. It seems we have to choose between either being consistent with DFS, or being consistent with POSIX.
I imagine this will also come up when we formally centralize these wrapper functions into DFS somehow.

dsikich replied:
@daltonbohning yeah I guess my thinking was I can try and make it consistent with DFS since these functions are going to be put with DFS, but it doesn't really matter to me as long as we agree on one way

daltonbohning replied:
@dsikich Maybe we can let it be for now, as long as the error handling is in place to begin with. And then we can formally decide as part of [DAOS-6098](https://jira.hpdd.intel.com/browse/DAOS-6098)

dsikich commented on src/utils/daos_hdlr.c:
```diff
@@ -48,6 +49,37 @@
+#define NUM_DIRENTS 	24
+#define MAX_FILENAME 	256
+
+struct dfs_daos {
```
@daltonbohning rename to fs_copy_dirent?

daltonbohning replied:
@dsikich That would definitely make it more clear. And I think the main thing is that we shouldn't use the dfs_ prefix because that is reserved for the DFS API.

dsikich commented:
@daltonbohning @mchaarawi added a commit to enforce the daos:// prefix when a pool/cont is passed in, and added some more error checking noticed by Dalton. I think an open issue is still a decision on what to return from the file_ wrappers. The POSIX io calls return -1 on error and set errno, and the dfs_ calls return errno. For now, I am making sure to return the errno set by the POSIX call if there is an error, so that the file_ wrappers return an errno if an io call fails.

mchaarawi commented:
> @dsikich Mostly minor style and error checking, except for the --src and --dst format, which I would like @mchaarawi's input on.
> Specifically:
> `--src daos:<pool>/<cont>/<path>` // Should this work?
> `--src daos:/<pool>/<cont>/<path>` // Should this work?
> `--src daos://<pool>/<cont>/<path>` // Should this work?

we want something similar to DUNS: so only "daos://" for pool/uuid style.
you can check the src/client/dfs/duns.c to see how i enforce that.

as for the extra path, that's an additional feature that i think should work, but can be done in a different iteration. instead of copying the entire container, we copy a specific dir in that container? or that dir itself can be a pointer to another container? both can be valid use cases.

mchaarawi commented on src/utils/daos.c:
```diff
@@ -1389,11 +1362,7 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 		if (argc == 3) {
```
minor: since you did that you can just:
if (argc == 3 || strcmp(argv[3], "copy") == 0)
  FS_COPY_CMDS_HELP();

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1743,78 +1788,103 @@ file_open_dfs(file_dfs_t *file_dfs, char *file, int flags, ...)
+		}
 	} else {
-		rc = -1;
+		rc = 1;
```
this is confusing now.
-1 is a generic error, but what is 1?
1 (errno speaking) is EPERM.
so you have to make up your mind on what you are returning.
-1 for error, or a positive errno in which case this needs to be EINVAL.

dsikich replied:
@mchaarawi updated to EINVAL

daltonbohning commented:
> we want something similar to DUNS: so only "daos://" for pool/cont style.
> you can check the src/client/dfs/duns.c to see how i enforce that.

Ok, great. I think @dsikich did update to enforce "daos://" exactly.

> as for the extra path, that's an additional feature that i think should work, but can be done in a different iteration. instead of copying the entire container, we copy a specific dir in that container? 

Yes, the path is a path in the container, which allows copying a subset of the container.

> ... or that dir itself can be a pointer to another container?

I don't believe this is supported here, or even in mpifileutils. A dir inside a container is treated as only a dir. I'll document this, so we can add support/tests in the future, both for this tool and mpifileutils.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1743,78 +1788,103 @@ file_open_dfs(file_dfs_t *file_dfs, char *file, int flags, ...)
+		rc = dfs_lookup(file_dfs->dfs, dname, O_RDWR, &parent, NULL, NULL);
+		if (parent == NULL) {
+			fprintf(stderr, "dfs_lookup %s failed \n", dname);
+			D_GOTO(out, rc = 1);
```
again, rc = 1 is confusing. -1 or the proper error code is better.

mchaarawi commented on src/utils/daos_hdlr.c:
```diff
@@ -1937,122 +2027,140 @@ file_lstat_dfs(file_dfs_t *file_dfs, const char *path, struct stat *buf)
 	} else {
-		got_size = -1;
-		fprintf(stderr, "File type not known: %s type=%d\n", file, file_dfs->type);
+		got_size = 1;
```
this is definitely wrong. setting got_size = 1 means you read 1 byte

daltonbohning commented on src/utils/daos.c:
```diff
@@ -1359,10 +1359,8 @@ help_hdlr(int argc, char *argv[], struct cmd_args_s *ap)
 	} else if (strcmp(argv[2], "filesystem") == 0 || strcmp(argv[2], "fs") == 0) {
-		if (argc == 3) {
+		if (argc == 3 || strcmp(argv[3], "copy") == 0) {
 			ALL_FS_CMDS_HELP();
```
@mchaarawi @dsikich Actually, I don't think we can do this, because then we don't get to see the --src and --dst options for copy.
Maybe?
```suggestion
		if (argc == 3 || strcmp(argv[3], "copy") == 0) {
			FS_COPY_CMDS_HELP();
```

dsikich replied:
@daltonbohning updated.

daltonbohning reviewed (commented):
@dsikich Just one minor change.
@mchaarawi Overall, I think this is decent. For what it's worth, it's also passing some basic functional tests (local, not yet in CI).
Do you think this is good enough to merge? Once this is merged, I can start formally introducing functional tests, which should weed out any issues.

daltonbohning reviewed (approved):
I think this is a good first pass. I have a local branch for testing basic functionality, which is working. Once this is merged, we can add those test cases to CI and work on more extensive testing.

mchaarawi commented:
actually i just realized that i don't think we can merge this PR since you created the PR from a branch in your repo.
So CI is not going to run any checkpatch / build against your changes.
you have to create a branch on daos-stack and a PR from that.

mchaarawi commented:
> actually i just realized that we cannot merge this PR since you created the PR from a branch in your repo.
> So CI is not going to run any checkpatch / build against your changes.
> you have to create a branch on daos-stack and a PR from that.

i guess this explains why checkpatch was not complaining on all the formatting issues that i spotted. there are probably others that i had missed.

dsikich commented:
> > actually i just realized that we cannot merge this PR since you created the PR from a branch in your repo.
> > So CI is not going to run any checkpatch / build against your changes.
> > you have to create a branch on daos-stack and a PR from that.
> 
> i guess this explains why checkpatch was not complaining on all the formatting issues that i spotted. there are probably others that i had missed.

@mchaarawi sure, I can do that. I will close this one and open a new PR.
